### ISSUE TITLE
Reject invalid linking

### ISSUE THREAD

**Issue by lode** [jelhan/json-api-bulk-create-extension#3]
The specification explains this exception case:

> Note: If two resources, which should be created, both have a to-one relationship to the same resource, the one listed in the document last wins. A server may validate the document against such scenarios and reject the request.

I was wondering why you chose for the option to let the last resource win?

I would opt for specifying that such request documents should be rejected. It feels a bit out of sync with how the document should be processed. Also I can imagine for different use cases the last or first one feels more logical. So being very clear about this would help I think.

**Comment by jelhan** [comment]
This is a consequence of the processing order:

> A server MUST create the resources in the order they appear in bulk:data and bulk:included. A server MUST create the resources in bulk:data before the resources in bulk:included.

As resources are processed in order, the last one wins. That's not a design decision. It's a consequence of a design decision.

The extension clearly defines the result of this edge case. Validation against it is _not_ needed to prevent an undefined result.

Validation has somw trade-offs in this case. Most noticeable: 1) Complexity implementing it and 2) performance. I feel it should be an implementation-specific decision if it's worth it.

**Comment by lode** [comment]
To me it isn't clear that it is an effect. From a client perspective I could just as well assume the other way around.

I would say that a client who sends such a request is bugged. And then they'll have a hard time finding the bug because nothing fails.

I think nothing blocks the specification from stating this MUST be a validation error. And indeed it is easy to implement as well.

**Comment by jelhan** [comment]
I agree that validating against that case would help spotting issues in clients early. That would be the main benefit.

But not sure how easy implementation would be for APIs supporting bulk creation with large payloads (e.g. for importing data from external sources). For those cases an API may want to avoid holding entire payload in memory and stream it instead. That complicates validation which requires knowledge about entire request document.

**Comment by lode** [comment]
True.

Isn't holding the entire payload anyway something the server needs to do?

> A server MUST perform the request atomically. If creating one resource fails, no resource MUST be created.

**Comment by jelhan** [comment]
> Isn't holding the entire payload anyway something the server needs to do?
> 
> > A server MUST perform the request atomically. If creating one resource fails, no resource MUST be created.

Not necessarily. Most databases support transactions.

**Comment by lode** [comment]
Yeah. I notice I'm too much in the context of a server first validating then creating.

Related to https://github.com/jelhan/json-api-bulk-create-extension/issues/2, I'm wondering whether the large import / streaming use case is something that adds complexity.
